File: C++/프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.cpp
#include <string>
#include <vector>
#include <cmath>
// #include <iostream>

using namespace std;

string findFn(long long number) {
    string res = "";
    
    while(number > 1) {
        res += to_string(number % 2);
        number /= 2;
    }
    
    return res + to_string(number);
}

vector<long long> solution(vector<long long> numbers) {
    vector<long long> answer;
    string binStr;
    int i;
    
    for(auto& number : numbers) {
        binStr = findFn(number);
        // cout << number << "의 이진수는 " << binStr << endl;
        if(binStr.find('0') != -1) {
            for(i = 0; i < binStr.size(); i++) { //제일 작은 0을 1로 변환
                if(binStr[i] == '0') {
                    number += pow(2, i);
                    break;
                }
            }
            
            for(int j = i; j >= 0; j--) {
                if(binStr[j] == '1') {
                    number -= pow(2, j);
                    break;
                }
            }
        } else {
            number -= pow(2, binStr.size() - 1);
            number += pow(2, binStr.size());
        }
        answer.push_back(number);
    }
    
    return answer;
}

File: Python3/프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.py
def solution(numbers):
    answer, fn_map = [], dict()
    bin_str, idx = "", 0
    
    for number in numbers:
        if number not in fn_map:
            bin_str = bin(number)[2:][::-1]
            if '0' in bin_str: # 0이 하나라도 존재할 경우
                idx = bin_str.index('0')
                for i in range(idx - 1, -1, -1):
                    if bin_str[i] == '1':
                        number -= 2 ** i
                        break
                answer.append(number + 2 ** bin_str.index('0'))
            else:
                number += 2 ** len(bin_str)
                number -= 2 ** bin_str.rfind('1')
                answer.append(number)
    return answer